text/0004-type-deduction.md

```diff
@@ -0,0 +1,214 @@
+- Feature Name: type_deduction
+- Start Date: 2020-10-13
+- RFC PR: [nushell/rfcs#0004](https://github.com/nushell/rfcs/pull/4)
+- Nushell Issue: [nushell/nushell#0000](https://github.com/nushell/nushell/issues/0000)
+
+# Summary
+
+[summary]: #summary
+
+The purpose of this RFC is to explore how type deduction can be implemented for nushell. Type deduction is helpful for e.G. aliases with typed variables, static analysis...
+
+# Motivation
+
+[motivation]: #motivation
+
+Currently no module for the sake of type deduction is implemented. Groundwork for such a module has been layed out in alias.rs. However, the code in alias.rs doesn't handle all cases.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Given a set of variables and a block of code, the purpose of the type deduction module (TDM) is to infer all possible types a variable can have so that the block of code is still valid. In nushell a "type" is represented by a SyntaxShape variant.
+The TDM infers by looking in which kind of expression (following: containing expression) a variable is used. Then it checks which expressions (following: sub expressions) are allowed inside the containing expression at the position of the variable and maps them to their corresponding SyntaxShape variant.
+For most cases it is sufficient to look at the nushell grammar, to figure out the allowed sub-expressions. E.G. `echo 1..$var` $var is of type Int as only Int is allowed in a range expression.
+Special cases are listed below.
+
+## As a commands mandatory positional argument | As a argument to flag 
+```shell Example positional argument
+ls $path | where $filter
+$path -> SyntaxShape::FilePath 
+$filter -> SyntaxShape::Math
+```
+```shell Example named argument
+cal --full-year $year
+$year -> SyntaxShape::Int
+```
+
+The shape of a variable used as a positional argument can be infered from the command signature.
+If the signature of the command is not available, no inference will be done.
+
+## As a commands optional positional argument
+```shell Variable in fixed position
+reminder: git log signature: git log [<options>] [<revision range>] [[--] <path>...]
+git log $arg ./src/
+```
+One can infer that $arg has to be a revision range, as there is no other possibility.
+
+```shell Variable in ambiguous position
+git log $arg
+```
+One can infer that $arg is either a revision range or a file path.
+
+```shell Variable in ambiguous position and dependencies
+cmd signature: cmd [<FilePath>] [<FileSize>] [<Block>] [<Int>...]
+cmd $a1 $a2
+```
+This situation is quite complex. For $a1 one can infer that it has to be of type FilePath, FileSize, Block or Int. The correct deduction for $a2 depends on the type of $a1. For example: If $a1 is Int, $a2 must also be Int.
```

Comment:
Let me clarify: The variables are also deduced in order: $a1 can be <opt_arg>, but then $a2 can't be anything previous of the <opt_arg>.

Should the parser work like it currently does? Is it intended behaviour that $a1 has to match FilePath here? Or is it rather a lack of maturity? 
If it is, I am strongly concerned about this design decission. 

Letting type deduction deviate from the parser internals is not a long term goal :) Thanks for pointing this one out.

Reply:
> Letting type deduction deviate from the parser internals is not a long term goal :)

Agreed. These two deviating would be very confusing 👍 

I'd like to have @jonathandturner speak more on this topic, but I'll still offer my thoughts. I'm thinking both what we have currently and what you're suggesting both have some room for being confusing. From the user's perspective though, I think what we have currently is far less confusing and far less surprising.

In your proposal, type inference is actually very similar to C++'s "substitution failure is not an error" (SFINAE) approach to generics. Essentially, you expand the signature `int, range, string` (all optional) into:

| # args | possible options |
|--------|------------------|
| 1 | `(int)`, `(range)`, `(string)` |
| 2 | `(int, range)`, `(int, string)`, `(range, string)` |
| 3 | `(int, range, string)` |

and take the first that would parse (based on the number of arguments provided). The part that I think is confusing with this approach is that the combinatorial explosion can be detrimental to command comprehension. Also, arguments are no longer positional, in the sense that if a user supplies an argument in position 1, it may apply to the signature argument in position 3. I also think it could encourage poor command design, but I don't have a good argument to tie to that thought right now.

With the current approach we basically expand into:

| # args | possible options |
|--------|------------------|
| 1 | `(int)` |
| 2 | `(int, range)` |
| 3 | `(int, range, string)` |

So you can see there's a lot of simplicity there. It's also very much like most programming languages (where you have optional arguments). Take a Ruby example:

```ruby
def foo(int=1, range=2..3, string="4"); end

foo(4..5) # 4..5 would fill in the `int` argument
```

I think we should go down a route that encourages as few positional arguments as possible. This is how I encourage people to write code too. Strive for a single positional argument, but if you need more make them keyword arguments (in nushell parlance, flags and/or subcommands).

Reply:
I didn't even thought about your approach towards this. I thought the support for optional arguments just wasn't there. Good to clarify that.

I didn't thought about this approach as it wouldn't support even common commands like 'git log' completly. Not to mention commands with more chained optional arguments out there. Or am I missing how nushell wants to support 
```shell
git log $user_supplied_path
```
in scripts later on?

My approach towards this problem wasn't inspiried by c++ SFINAE, but by viewing signatures as a subset of regex (signatures only support optional and * at the end).

> The part that I think is confusing with this approach is that the combinatorial explosion can be detrimental to command comprehension.

Thanks for pointing out in a crystal clear manner that the 2 variable have to fit a combination of the optional arguments. I didn't noticed :). Maybe it helps to simplify things.

I don't see a problem of combinatorial explosion. If one writes:
```shell
git log $path
```
for me as a reader, its clear what $path would fit into. This problem might be handled with good variable naming?


> I think we should go down a route that encourages as few positional arguments as possible. This is how I encourage people to write code too. Strive for a single positional argument, but if you need more make them keyword arguments (in nushell parlance, flags and/or subcommands).

Totaly agree.
But from my perspective any shell should try it's best to support existing commands fully. Thats how I understand nushell goal #2 too.

The best way forward might be to make a new rfc for this decision? I will update the rfc / pr code to respect the current implementation.